src/MIDebugEngine/Engine.Impl/BreakpointManager.cs

```diff
@@ -141,29 +141,66 @@ private AD7PendingBreakpoint BindToAddress(string bkptno, ulong addr, /*OPTIONAL
             return pending;
         }
 
-        public AD7BoundBreakpoint FindHitBreakpoint(string bkptno, ulong addr, /*OPTIONAL*/ TupleValue frame, out bool fContinue)
+        // Return all bound breakpoints bound to an address regardless of bptno
+        AD7BoundBreakpoint[] FindBoundBreakpointsAtAddress(ulong addr)
+        {
+            lock (_pendingBreakpoints)
+            {
+                List<AD7BoundBreakpoint> matchingBoundBreakpoints = new List<AD7BoundBreakpoint>();
+                foreach (AD7PendingBreakpoint currPending in _pendingBreakpoints)
+                {
+                    matchingBoundBreakpoints.AddRange(Array.FindAll(currPending.EnumBoundBreakpoints(), (b) => b.Addr == addr));
+                }
+
+                return matchingBoundBreakpoints.ToArray();
+            }
+        }
+
+
+        public AD7BoundBreakpoint[] FindHitBreakpoints(string bkptno, ulong addr, /*OPTIONAL*/ TupleValue frame, out bool fContinue)
         {
             fContinue = false;
+            List<AD7BoundBreakpoint> hitBoundBreakpoints = new List<AD7BoundBreakpoint>();
             lock (_pendingBreakpoints)
             {
-                AD7BoundBreakpoint bbp;
-                AD7PendingBreakpoint pending = BindToAddress(bkptno, addr, frame, out bbp);
-                if (pending == null)
+                AD7BoundBreakpoint[] hitBps;
+
+                // For the case where an address is provided for the breakpoint, match based on address.
+                // This will cause the engine to send a breakpoint event accounting for all breakpoints
+                // that are hit. 
+                // However, clrdbg does not support addresses on its breakpoints, so the bkptno code path is
+                // preserved.
+                if (addr != 0)
                 {
-                    return null;
+                    hitBps = FindBoundBreakpointsAtAddress(addr);
                 }
-                fContinue = true;
-                if (!pending.Enabled || pending.Deleted || pending.PendingDelete)
+                else
```

Comment:
Do we need to lookup by breakpoint number in the case that we previously didn't know the address of the breakpoint? I think that can happen if the breakpoint binds to multiple addresses.